### PR TITLE
Dispose drawables that are not yet loaded inside CompositeDrawable

### PR DESCRIPTION
**Pull request by nyquillerium**
Fixes https://github.com/ppy/osu/issues/5120

Previously, `Drawable`s that are being loaded asynchronously will not get disposed if the `CompositeDrawable` thats loading it gets disposed and if the `Drawable` thats being loaded hasn't been added to the scene yet. 

This change makes it so every `Drawable` that gets loaded asynchronously by a `CompositeDrawable` will have `Dispose` called on them if they aren't of the `Loaded` state yet.

### CODE REVIEW

**Comment by peppy** [comment]
95% sure the test failure is a result of your changes. may want to test that one locally.

**Comment by nyquillerium** [comment]
Yeah, that one was caused by my changes. I fixed it locally but I'm now running into the same failure on https://ci.appveyor.com/project/peppy/osu-framework/builds/25928295. Not sure if that one would be too.

**Comment by peppy** [comment]
They are different.

**Comment by peppy** [comment]
You're not removing the components after they are added. There's no reason to keep a reference (even a weak one) to them. It's taking up extra memory and could potentially add up over the lifespan of a container).

File: osu.Framework/Graphics/Containers/CompositeDrawable.cs [hunk 1]
```diff
@@ -262,6 +270,10 @@ protected override void Dispose(bool isDisposing)
 
             InternalChildren?.ForEach(c => c.Dispose());
 
+            foreach (var d in loadingComponents)
+                if (!d.IsLoaded)
+                    d.Dispose();
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Containers/CompositeDrawable.cs]: Can this be done on a cancelled/faulted-only continuation of the load task instead, removing the requirement of the list?

Something like`.ContinueWith(() => { /*dispose */}, TaskContinuationOptions.NotOnRanToCompletion)`

File: osu.Framework/Graphics/Containers/CompositeDrawable.cs [hunk 2]
```diff
@@ -138,6 +140,21 @@ protected internal Task LoadComponentAsync<TLoadable>(TLoadable component, Actio
...
+                loadingComponents = new WeakList<Drawable>();
+
+            foreach (var d in components)
+            {
+                loadingComponents.Add(d);
+                d.OnLoadComplete += _ =>
+                {
+                    loadingComponents.Remove(d);
+
+                    if (!loadingComponents.Any())
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/CompositeDrawable.cs]: this isn't required

File: osu.Framework/Graphics/Containers/CompositeDrawable.cs [hunk 3]
```diff
@@ -262,6 +273,10 @@ protected override void Dispose(bool isDisposing)
 
             InternalChildren?.ForEach(c => c.Dispose());
 
+            foreach (var d in loadingComponents ?? Enumerable.Empty<Drawable>())
+                if (!d.IsLoaded)
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/CompositeDrawable.cs]: why is this needed?

Comment (nyquillerium) [reply osu.Framework/Graphics/Containers/CompositeDrawable.cs]: This should be safe to remove
